kotlinx-coroutines-core/common/src/flow/operators/Chunk.kt

```diff
@@ -0,0 +1,211 @@
+
+        /**
+         * Collects upstream into a buffer and emits its content as a list, when specified size is reached.
+         * This implementation is purely sequential. If concurrent upstream collection and downstream emissions are
+         * desired, one can use a buffer() operator after chunking
+         *
+         * @param size Exact size of emitted chunks. Only the last emission may be smaller.
+         */
+        @Suppress("FunctionName")
+        public fun BySize(size: Int): ChunkingMethod = SizeBased(size)
+    }
+}
+
+private class NaturalChunking(private val maxSize: Int) : ChunkingMethod {
+
+    init {
+        requirePositive(maxSize)
+    }
+
+    override fun <T> Flow<T>.chunk(): Flow<List<T>> = scopedFlow { downstream ->
+        val upstream = buffer(maxSize).produceIn(this)
+
+        while (!upstream.isClosedForReceive) {
+            val chunk = upstream.awaitFirstAndDrain(maxSize)
+            if (chunk.isNotEmpty()) downstream.emit(chunk)
+        }
+    }
+}
+
+private class TimeBased(private val intervalMs: Long, private val maxSize: Int) : ChunkingMethod {
+
+    init {
+        requirePositive(intervalMs)
+        requirePositive(maxSize)
+    }
+
+    override fun <T> Flow<T>.chunk(): Flow<List<T>> = scopedFlow { downstream ->
+        val upstreamCollection = Job()
+        val upstream = produce<T>(capacity = maxSize) {
+            collect { element -> channel.send(element) }
+            upstreamCollection.complete()
+        }
+
+        whileSelect {
+            upstreamCollection.onJoin {
+                val chunk = upstream.drain(maxElements = maxSize)
+                if (chunk.isNotEmpty()) downstream.emit(chunk)
+                false
+            }
+
+            onTimeout(intervalMs) {
+                val chunk = upstream.drain(maxElements = maxSize)
+                if (chunk.isNotEmpty()) downstream.emit(chunk)
+                true
+            }
+        }
+    }
+}
+
+private class SizeBased(private val size: Int) : ChunkingMethod {
+
+    init {
+        requirePositive(size)
+    }
+
+    override fun <T> Flow<T>.chunk(): Flow<List<T>> = flow {
+        val accumulator = ArrayList<T>(size)
+        collect { element ->
+            accumulator.add(element)
+            if (accumulator.size == size) emit(accumulator.drain())
+        }
+        if (accumulator.isNotEmpty()) emit(accumulator)
+    }
+}
+
+private class TimeOrSizeBased(private val intervalMs: Long, private val maxSize: Int) : ChunkingMethod {
+
+    init {
+        requirePositive(intervalMs)
+        requirePositive(maxSize)
+    }
+
+    override fun <T> Flow<T>.chunk(): Flow<List<T>> = scopedFlow { downstream ->
+        val emitNowAndMaybeContinue = Channel<Boolean>(capacity = Channel.RENDEZVOUS)
+        val elements = produce<T>(capacity = maxSize) {
+            collect { element ->
+                val hasCapacity = channel.trySend(element).isSuccess
+                if (!hasCapacity) {
```

Comment:
when maxSize is 1 , hasCapacity returns true so  emitNowAndMaybeContinue cannot trigger